### ISSUE TITLE
CustomConfig List >5k Items

### ISSUE THREAD

**Issue by gudatta** [pnp/custom-learning-office-365#434]
### What is the context of your question

- [ ] Installation
- [ ] Learning Pathways Content
- [ ] SharePoint Success Content
- [ ] Administration
- [ ] Custom Content
- [X] Viewer Web Part

### What is your question

Our customer recently started getting errors because it appears a new item was added to the CustomConfig sharepoint list which pushed it over the 5000 list view item threshold. This issue is causing most of the content on the LP site to stop loading.

From the research we’ve done so far we were able to deduce that it appears the dataservice.ts file is trying to read data from the cache but is unable to and just recursively adds new items to the CustomConfig list each time the code is executed which has caused the list to balloon and effectively breaks the entire Learning Pathways site.

#### Updates since 12/7?
Have there been any recent updates -around 12/7/2020, that would have caused these issues?

#### Delete List & Recreate?
What is my recourse if the CustomCDN list item was deleted? Can I add it back manually? If so, what values should I use?

**Comment by juliemturner** [comment]
Usually list items that are deleted in the UI are still in the recycle bin and can be restored.  I'd check that first.  If not then your only recourse is to add the CDN's again. The only way that becomes tricky is if you have customizations on those other CDN's (i.e. have custom playlists/assets, etc). If you use the same GUIDs when you add them through the UI then it should fix the problem. To get the GUIDs you can look in the CustomPlaylists list for instance and find the instances of the GUIDs

![image](https://user-images.githubusercontent.com/7570936/103777099-646bb580-4ffe-11eb-8467-0e97781148b1.png)

**Comment by godlikesme** [comment]
I don't see any GUIDS in the CDN field. They all just say Default.

**Comment by juliemturner** [comment]
Then you didn't have any customizations on the CDN's you had added so you just add any back through the UI.

**Comment by godlikesme** [comment]
But I'm still getting errors all over the site. :(

<img width="825" alt="Screen Shot 2021-01-06 at 3 01 23 PM" src="https://user-images.githubusercontent.com/76570480/103814682-1bccf000-5030-11eb-863e-41f031c67651.png">

**Comment by juliemturner** [comment]
If you look at those error messages it implies that you have records in the CustomConfig with the title CustomSubcategories and CustomCDN but the field JSONData has no data. Not sure what you've done to accomplish this but I'm concerned after I suggested you remove the CustomConfig entries and then you asked me how to restore list items. At this point it's becoming pretty difficult for me to assist you. 

Let's start with the following questions:
- Are you sure you had custom content packs added to your configuration? If so what ones?
- Do you have any custom playlists or custom assets?
- Do you know if you created any custom sub categories?

**Comment by godlikesme** [comment]
But it does have content in JSONData. That's what confuses me. I made backup copies of the json before clearing out the list.
<img width="612" alt="Screen Shot 2021-01-06 at 3 55 52 PM" src="https://user-images.githubusercontent.com/76570480/103819109-cac0fa00-5037-11eb-8e09-33e9d41ef05b.png">

**Comment by juliemturner** [comment]
Delete the CustomCDN entry, it has no data and shouldn't be there.
You could have changed the formatting or put a space in that is confusing the other entry but start by deleting CustomCDN entry.
Also, based on this screen shot you're again with an issue that you have multiple CustomConfig entries... please make sure you're up to date with the latest web part version, this should not be happening.

**Comment by godlikesme** [comment]
I cleared everything out of the CustomConfig list and now the errors are gone. However, we are now right back where we started because each time I load **_ANY_** page in the site it adds another item to the CustomConfig list. I just don't know what could be causing this. When researching this issue with @gudatta last week we had discovered that it looked like the code in the dataservices.ts file was the issue as he mentioned in his initial post above. It would appear that after the code checks the cache, if it doesn't find the data it adds an item to the CustomConfig list but doesn't first check to see if the same entry already exists so it just keeps adding new ones each time the method in the dataservices.ts file is called.

**Comment by godlikesme** [comment]
The first time I load the CustomLearningAdmin.aspx page after clearing out all of the items in the CustomConfig list everything works fine and there are no errors on that page when I use the browser dev tools. However, each subsequent refresh of the page -or any other page on the site shows the same "unexpected token > in json at position 0" error. It's almost like the code that initially populates the CustomConfig list is actually inserting invalid json data. Which is very, very odd because how would this have worked all along if no one has changed anything?

**Comment by juliemturner** [comment]
I can guarantee you there is no bug in that code as it's working in thousands of tenants... but that said you could have a configuration issue.

- Can you validate what version of the solution you are using, if not 4.1.4 then please [update](https://github.com/pnp/custom-learning-office-365#updating-the-solution).
- Forgetting about the viewer page for the time being if you remove all the CustomConfig entries from the CustomConfig list and refresh the CustomLearningAdmin.aspx page multiple times does it also add new items?
- Please take a screen shot of your About page and share it with me
![image](https://user-images.githubusercontent.com/7570936/104194096-dffaa780-53ee-11eb-8bd4-590b68c78ccc.png)

**Comment by godlikesme** [comment]
- Version 3.0.0 (and yes, I am in the process of getting the app upgraded to the latest version)
- As I've mentioned above, even after completely clearing out the CustomConfig list so that it's totally empty, each time I refresh the CustomLearningAdmin.aspx page it adds another list item called "CustomConfig" into the CustomConfig list.
- I am not authorized to take a screenshot to post for public viewing. Where can I find the About page? I don't see any links for it.

With respect to the "bug" assertion, I completely understand that LP is running error free in thousands of environments -and our environment was one of those thousands of environments up until last month. My thoughts are that there had to be some kind of configuration change that would have caused this issue. There is no other logical explanation. 

With that being said, the fact that multiple CustomConfig list items continue to be added to the list does lead me to believe that there is something "off' with the method that handles the insertion of new items into that list. Now, it could be that through all the testing that went into the LP code, the exact conditions that revealed this bug never came up until now. It could also be that the issue would be fixed in the most recent version of the LP deployed but until I can get to that point I'm kinda stuck. The fact remains that these items continue to be added to the list unchecked which means that the current version of the package does not have logic in that method that should confirm that the CustomConfig list item entry does not yet exist before creating a new one.

Anyhow, I'll keep poking around and see if anything else jumps out at me and let you know when the latest version of the app has been deployed and what the outcome is. Thanks again for you help.

**Comment by juliemturner** [comment]
Oh... ok, full stop... if you're running that exact version (3.0.0) then I suspect that's your problem. What might have changes I cannot tell you as there have been no changes that could affect what you're talking about I suspect it's been doing this for some time you just happened to hit the list limit so now it's something you're recognizing. When you get to version 4 that should solve your problems. That said I'm not sure it's a good idea for you to try and go to version 4 when version 3 isn't working.  

What I don't have a clear picture of yet is if you have any custom playlists or assets and any custom sub-categories.  If you do then this becomes even more important to get fixed first.  I would suggest you try upgrading to the last stable version 3 of the webpart which was 3.0.4, you can get it here: https://github.com/pnp/custom-learning-office-365/tree/main/installation/Archive/v3_0_4

**Comment by godlikesme** [comment]
I just did the upgrade to 4.1.4 (after first trying 3.0.4) and the problem still persists. :(

**Comment by godlikesme** [comment]
I also ran the CDN update script.

**Comment by juliemturner** [comment]
You're on the CustomLearningAdmin.aspx page... and you went through the upgrade process?

**Comment by godlikesme** [comment]
Yup.

On Jan 13, 2021, at 10:44 AM, Julie Turner <notifications@github.com> wrote:

﻿

You're on the CustomLearningAdmin.aspx page... and you went through the upgrade process?

**Comment by juliemturner** [comment]
Then I need to see a screen shot of your about page.

**Comment by godlikesme** [comment]
I’ve been looking around and I’m not sure which “About” page you’re referring to.

On Jan 13, 2021, at 11:59 AM, Julie Turner <notifications@github.com> wrote:

﻿

Then I need to see a screen shot of your about page.

**Comment by juliemturner** [comment]
>- Please take a screen shot of your About page and share it with me
![image](https://user-images.githubusercontent.com/7570936/104194096-dffaa780-53ee-11eb-8bd4-590b68c78ccc.png)

**Comment by godlikesme** [comment]
I saw the screenshot earlier. My question is, where is this page? What is the URL?

On Jan 13, 2021, at 1:05 PM, Julie Turner <notifications@github.com> wrote:

﻿

  *   Please take a screen shot of your About page and share it with me
[image]<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F7570936%2F104194096-dffaa780-53ee-11eb-8bd4-590b68c78ccc.png&data=04%7C01%7C%7Ce5139c70c2294a4d82bb08d8b7eddaa6%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637461579485823824%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=CV1L0pB5TxukVOSfUcPNtu7Y%2BtucPnIeFQUCIPuWER0%3D&reserved=0>

**Comment by juliemturner** [comment]
It's the CustomLearningAdmin.aspx page. Sorry, I thought that was obvious from the heading of content packs.

**Comment by gudatta** [comment]
@godlikesme - Refer to this link https://docs.microsoft.com/en-us/office365/customlearning/custom_manualsetup#step-6--execute-powershell-configuration-script

**Comment by juliemturner** [comment]
@gudatta -- why would he do that? I really need the information in the About Web Part screen... basically he's got something not set up in a way that we support that is causing the system to not find the CustomConfig list item, I'm trying to determine what that is.

**Comment by gudatta** [comment]
@juliemturner - To help address the "Where is the 'Admin' page". Refer to that step of the process which created the page. That was the question I was trying to help address. Isn't the admin page where you get the "About" information?

**Comment by juliemturner** [comment]
Um.. yes, but since this entire thread has been about him telling me he's testing things using the admin page I assume he knows where it is...

**Comment by godlikesme** [comment]
![Uploading CustomLearningAdmin.png…]()

**Comment by juliemturner** [comment]
The issue is this "Unexpected token < in JSON at position 0" error...  Which means it's finding the record but can't read it.... which is why it's creating a new one... 

Ok, if you have the customsubcategories and customcdn list items backed up then just delete everything from the CustomConfig list and then refresh CustomLearningAdmin.aspx page and see if you still get the error.

**Comment by godlikesme** [comment]
I already told you (see previous post) that I did remove everything. Whatever json data is being inserted into the CustomConfig list is coming from the initial and subsequent refreshes of the CustomAdmin.aspx page.

**Comment by juliemturner** [comment]
I'm sorry but this entire configuration hasn't been communicated very clearly from the get go... l ask questions and you answer different ones so it's hard to follow. Please send me a screen shot of the values currently in teh CustomConfig list as well as the data in the JSONData field of one of the lists. Something is wrong with your configuration but I cannot seem to tease out what it is.

**Comment by godlikesme** [comment]
![CustomConfig](https://user-images.githubusercontent.com/76570480/104520163-a7162a80-55c8-11eb-9f44-417f9ea4b268.png)

[CustomConfigJsonData.txt](https://github.com/pnp/custom-learning-office-365/files/5811442/CustomConfigJsonData.txt)

**Comment by juliemturner** [comment]
I'm so lost... the dataset you sent me loaded just fine in my tenant... so there is clearly nothing wrong with the payload just the ability to read it again.

- Are you logged in as a site collection admin? If not can you use an account that is one instead just to take all permission type issues out of it and see if that solves it?
- Can you export the Network Log and the Console log to a text file and share them with me [here](https://sympraxis-my.sharepoint.com/:f:/g/personal/julie_turner_sympraxisconsulting_com/EvG2R7SlZ0pGm8AXUfO16AUBtdJS6ayKtoAyNmUr2HOXaA?e=7kne5I) - For directions on how to do that see this [document](https://github.com/pnp/custom-learning-office-365/blob/main/installation/DiagnosisPreCheck.md).

**Comment by juliemturner** [comment]
Just wanted to check in and see if you have fixed the issue yourself or if you still are looking for assistance?

**Comment by godlikesme** [comment]
I finally gave up, blew the site away and recreated it. I just couldn’t wait any longer. This whole experience has been totally bizarre. I sincerely do appreciate your help though. Thank you.

On Jan 28, 2021, at 10:31 AM, Julie Turner <notifications@github.com> wrote:

﻿

Just wanted to check in and see if you have fixed the issue yourself or if you still are looking for assistance?

**Comment by juliemturner** [comment]
I couldn't agree more... Ok, then i'll close the issue.
